File: day2/day2.js
const stdin = process.openStdin();
let data = '';
stdin.on('data', function (chunk) {
    data += chunk;
});

const points1 = {
    //rock
    A: {
        //rock
        X: 1 + 3,
        //paper
        Y: 2 + 6,
        //scissors
        Z: 3 + 0
    },
    //paper
    B: {
        X: 1 + 0,
        Y: 2 + 3,
        Z: 3 + 6
    },
    //scissors
    C: {
        X: 1 + 6,
        Y: 2 + 0,
        Z: 3 + 3
    },
}

const points2 = {
    //rock
    A: {
        //lose
        X: 3 + 0,
        //draw
        Y: 1 + 3,
        //win
        Z: 2 + 6
    },
    //paper
    B: {
        X: 1 + 0,
        Y: 2 + 3,
        Z: 3 + 6
    },
    //scissors
    C: {
        X: 2 + 0,
        Y: 3 + 3,
        Z: 1 + 6
    },
}

stdin.on('end', function () {
    const lines = data.split('\n').filter(x => !!x).map(line => line.split(' '));
    const results1 = lines.map(line => {
        const [you, me] = line;
        return { me, you, points: points1[you][me] }
    })
    console.log({ results1 });
    const result1 = results1.reduce((acc, curr) => acc + curr.points, 0);
    console.log('part 1', result1);

    const results2 = lines.map(line => {
        const [you, me] = line;
        return { me, you, points: points2[you][me] }
    })
    console.log({ results2 });
    const result2 = results2.reduce((acc, curr) => acc + curr.points, 0);
    console.log('part 2', result2);
})

File: day2/day2.exs
plays =
  IO.read(:stdio, :all)
  |> String.trim()
  |> String.split("\n")
  |> Enum.map(&String.split(&1, " "))

defmodule RockPaperScissors do
  def result(:rock, :scissors), do: :win
  def result(:paper, :rock), do: :win
  def result(:scissors, :paper), do: :win
  def result(x, x), do: :draw
  def result(_, _), do: :lose

  def points(:rock), do: 1
  def points(:paper), do: 2
  def points(:scissors), do: 3

  def points(me, opponent) do
    points(me) +
      case result(me, opponent) do
        :win -> 6
        :draw -> 3
        :lose -> 0
      end
  end

  def which(opponent, outcome) do
    [:rock, :paper, :scissors]
    |> Enum.find(fn me -> result(me, opponent) == outcome end)
  end
end

code1 = %{
  "A" => :rock,
  "B" => :paper,
  "C" => :scissors,
  "X" => :rock,
  "Y" => :paper,
  "Z" => :scissors
}

part1 =
  plays
  |> Enum.map(fn [l, r] ->
    opponent = Map.get(code1, l)
    me = Map.get(code1, r)

    RockPaperScissors.points(me, opponent)
  end)
  |> Enum.sum()

IO.puts("Part 1: #{part1}")

code2 = %{
  "A" => :rock,
  "B" => :paper,
  "C" => :scissors,
  "X" => :lose,
  "Y" => :draw,
  "Z" => :win
}

part2 =
  plays
  |> Enum.map(fn [l, r] ->
    opponent = Map.get(code2, l)
    outcome = Map.get(code2, r)
    me = RockPaperScissors.which(opponent, outcome)

    RockPaperScissors.points(me, opponent)
  end)
  |> Enum.sum()

IO.puts("Part 2: #{part2}")
